src/Dialect/ONNX/ElementsAttr/Strides.hpp

```diff
@@ -119,15 +120,122 @@ struct StridedArrayRef : public llvm::ArrayRef<T> {
       : Base(array), strides(strides) {}
 };
 
-template <typename Iterator, typename... Args,
-    typename Action = llvm::function_ref<void(Iterator, const Args *...)>>
-Iterator traverseStrides(llvm::ArrayRef<int64_t> shape, Iterator begin,
-    StridedArrayRef<Args>... src, Action &&act);
+// Suppose strided arrays {{array1,strides1},...,{arrayN,stridesN}} all
+// represent the same shape, then
+// it = StridesIterator<N>(shape, {strides1,...,stridesN}) iterates over the
+// shape's elements in row-major order, with it->index describing the shape's
+// corresponding multidimensional index, it->flattenedIndex the row-major order
+// flattened linear index, and it->pos is an array of positions pos1,...,posN
+// in array1,...,arrayN that represent the index. i.e.,
+// it->pos1 == getStridesPosition(it->index, strides1) etc.
+//
+// For example, this can be used to compare the contents of two strided arrays
+// with the same shape:
+//
+//   template <typename T, typename StridedArray = StridedArrayRef<T>>
+//   bool equal(ArrayRef<int64_t> shape, StridedArray lhs, StridedArray rhs) {
+//     StridesIterator<2> begin(shape, {lhs.strides, rhs.strides}), end(shape);
+//     return llvm::all_of(llvm::make_range(begin, end),
+//         [](const auto &it) { return src0[it.pos[0]] == src1[it.pos[1]]; })
+//   }
+//
+template <size_t N>
+class StridesIterator {
+  struct value_type {
+    value_type(unsigned rank, uint64_t flattenedIndex = 0)
+        : pos{}, flattenedIndex(flattenedIndex), index(rank, 0) {}
+    std::array<size_t, N> pos;
+    uint64_t flattenedIndex;
+    llvm::SmallVector<uint64_t, 6> index;
+  };
+  using difference_type = int64_t;
+  using pointer = const value_type *;
+  using reference = const value_type &;
+  using iterator_category = std::forward_iterator_tag;
+
+  const llvm::ArrayRef<int64_t> shape;
+  const std::array<llvm::ArrayRef<int64_t>, N> strides;
+  value_type value;
+
+public:
+  // Begin iterator.
+  StridesIterator(llvm::ArrayRef<int64_t> shape,
+      std::array<llvm::ArrayRef<int64_t>, N> strides)
+      : shape(shape), strides(strides), value(shape.size()) {
+    for (auto dim : shape)
+      assert(dim > 0 && "shape must describe non-empty tensor");
+    for (unsigned i = 0; i < N; ++i)
+      assert(shape.size() == strides[i].size() && "shape, strides mismatch");
+  }
+
+  // End iterator: ends after the given number of iterations.
+  StridesIterator(size_t iterations) : value{0, iterations} {}
+
+  // End iterator: ends after one traversal of shape.
+  StridesIterator(llvm::ArrayRef<int64_t> shape)
+      : StridesIterator(mlir::ShapedType::getNumElements(shape)) {}
+
+  StridesIterator(const StridesIterator &) = default;
```

Comment:
I'd be interested to know if these definitions are needed or if they will be auto generated by the compiler

Reply:
they are not, I added the comment:
```
  // These declarations are redundant, the compiler generates this constructor
  // and operator automatically, but included to flag that StridesIterator can
  // be copied, which is used in the implementation of operator++(int) below.
```